lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,202 @@
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of BitCast to nonstandard types. LLVM
+// transformations bitcast some vector types to scalar types, which are not
+// universally supported across all targets. We need ensure that "optimized"
+// LLVM IR doesn't have primitive types other than supported by the
+// SPIR target (i.e. "scalar 8/16/32/64-bit integer and 16/32/64-bit floating
+// point types, 2/3/4/8/16-element vector of scalar types").
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spv-lower-bitcast-to-nonstandard-type"
+
+#include "SPIRVInternal.h"
+
+#include "llvm/IR/PassManager.h"
+
+#include <stack>
+#include <utility>
+
+using namespace llvm;
+
+namespace SPIRV {
+
+class SPIRVLowerBitCastToNonStandardTypePass {
+public:
+  SPIRVLowerBitCastToNonStandardTypePass() {}
+
+  PreservedAnalyses
+  runLowerBitCastToNonStandardType(Function &F, FunctionAnalysisManager &FAM) {
+    // This pass doesn't cover all possible uses of non-standard types, only
+    // known. We assume that bad type won't be passed to a function as
+    // parameter, since it added by an optimization.
+    bool Changed = false;
+
+    std::vector<std::pair<Instruction *, VectorType *>> BCastsToNonStdVec;
+    std::stack<Instruction *> InstsToErase;
+    for (auto &BB : F)
+      for (auto &I : BB)
+        if (auto *BC = dyn_cast<BitCastInst>(&I)) {
+          auto *DestTy = BC->getDestTy();
+          if (auto *ElemTy = dyn_cast<PointerType>(DestTy))
+            DestTy = ElemTy->getElementType();
+          if (auto *DestVecTy = dyn_cast<VectorType>(DestTy)) {
+            uint64_t NumElemsInDestVec =
+                DestVecTy->getElementCount().getValue();
+            if (NumElemsInDestVec != 2 && NumElemsInDestVec != 3 &&
+                NumElemsInDestVec != 4 && NumElemsInDestVec != 8 &&
+                NumElemsInDestVec != 16)
+              BCastsToNonStdVec.push_back(std::make_pair(&I, DestVecTy));
+          }
+        }
+
+    for (auto &I : BCastsToNonStdVec)
+      Changed |= lowerBitCastToNonStdVec(I, InstsToErase);
+
+    while (!InstsToErase.empty()) {
+      InstsToErase.top()->eraseFromParent();
+      InstsToErase.pop();
+    }
+
+    return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+  }
+
+  bool lowerBitCastToNonStdVec(std::pair<Instruction *, VectorType *> &Inst,
+                               std::stack<Instruction *> &InstsToErase) {
+    bool Changed = false;
+    std::vector<Instruction *> InstsToInsert;
+    Instruction *I = Inst.first;
+    VectorType *DestVecTy = Inst.second;
+    uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+    InstsToErase.push(I);
+    AddrSpaceCastInst *Src = nullptr;
+    auto *ASCastInstIter = I;
+    while ((ASCastInstIter = ASCastInstIter->getNextNode())) {
```

Comment:
It would be nice to have comment about a purpose of blocks started at lines 106, 128, 138.